lib/collection.js

```diff
@@ -340,9 +323,59 @@ Collection.prototype.find = deprecateOptions(
       selector = { _id: selector };
     }
 
-    if (!options) options = {};
+    const findSchema = {
+      limit: { type: 'number', default: 0 },
+      sort: { type: ['array', 'object'] },
+      projection: { type: 'object' },
+      fields: { type: 'object' },
+      skip: { type: 'number', default: 0 },
+      hint: { type: ['string', 'object'], default: this.s.collectionHint },
+      explain: { type: 'boolean' },
+      snapshot: { type: 'boolean' },
+      timeout: { type: 'boolean' },
+      tailable: { type: 'boolean' },
+      batchSize: { type: 'number' },
+      returnKey: { type: 'boolean' },
+      maxScan: { type: 'number' },
+      min: { type: 'number' },
+      max: { type: 'number' },
+      showDiskLoc: { type: 'boolean' },
+      showRecordId: { type: 'boolean' },
+      comment: { type: 'string' },
+      raw: { type: 'boolean', default: this.s.raw },
+      promoteLongs: { type: 'boolean', default: this.s.promoteLongs },
+      promoteValues: { type: 'boolean', default: this.s.promoteValues },
+      promoteBuffers: { type: 'boolean', default: this.s.promoteBuffers },
+      readPreference: { type: [ReadPreference, 'object', 'string'] },
+      partial: { type: 'boolean' },
+      maxTimeMS: { type: 'number' },
+      collation: { type: 'object' },
+      session: { type: ClientSession },
+      promiseLibrary: { rejectUser: true },
+      slaveOk: { type: 'boolean', default: this.s.db.slaveOk },
+      awaitData: { rejectUser: true },
+      noCursorTimeout: { type: 'boolean', rejectUser: true },
+      ignoreUndefined: { rejectUser: true }
+    };
+
+    const finalOptions = validate(
+      findSchema,
+      options,
+      {
+        readPreference: resolveReadPreference(options, { db: this.s.db, collection: this }),
+        hint:
+          options && options.hint != null
+            ? normalizeHintField(options.hint)
+            : this.s.collectionHint,
+        noCursorTimeout: options ? options.timeout : null,
+        awaitData: options ? options.awaitdata : null,
+        promiseLibrary: this.s.promiseLibrary,
+        ignoreUndefined: this.s.options.ignoreUndefined
+      },
+      { validationLevel: 'error' }
```

Comment:
Let's default to `warn` for the first iteration here, as you said earlier people might accidentally be depending on broken options and we want to give them a chance to fix that rather than breaking their application.

Also, I would recommend setting this to a constant called like `LOG_DEFAULT` at the top of the file, so you can easily toggle between error modes during testing.